cpp/src/arrow/io/memory.cc

```diff
@@ -82,20 +89,19 @@ Status BufferOutputStream::Close() {
 
 bool BufferOutputStream::closed() const { return !is_open_; }
 
-Status BufferOutputStream::Finish(std::shared_ptr<Buffer>* result) {
+Result<std::shared_ptr<Buffer>> BufferOutputStream::Finish() {
```

Comment:
At some point, it might be advantageous to mark this as rvalue qualified:
```suggestion
Result<std::shared_ptr<Buffer>> BufferOutputStream::Finish() && {
```

Reply:
Hmm, why that?

Reply:
We'd get compiler warnings (use after move) if a BufferOutputStream were used after Finish has been called